test/ambiguous.jl

```diff
@@ -0,0 +1,59 @@
+# DO NOT CHANGE LINE NUMBERS BELOW
+ambig(x, y) = 1
+ambig(x::Integer, y) = 2
+ambig(x, y::Integer) = 3
+ambig(x::Int, y::Int) = 4
+ambig(x::Number, y) = 5
+# END OF LINE NUMBER SENSITIVITY
+
+ambigs = Any[[], [3], [2,5], [], [3]]
+
+mt = methods(ambig)
+
+getline(m::Method) = m.line - 1  # -1 for the comment at the top
+
+for m in mt
+    ln = getline(m)
+    atarget = ambigs[ln]
+    if isempty(atarget)
+        @test m.ambig == nothing
+    else
+        aln = Int[getline(a) for a in m.ambig]
+        @test sort(aln) == atarget
+    end
+end
+
+@test ambig("hi", "there") == 1
+@test ambig(3.1, 3.2) == 5
+@test ambig(3, 4) == 4
+@test_throws AmbiguousError ambig(0x03, 4)
+@test_throws AmbiguousError ambig(0x03, 4)  # test that not inserted into cache
+
+# Ensure it still works with potential inlining
+callambig(x, y) = ambig(x, y)
+@test_throws AmbiguousError callambig(0x03, 4)
+
+## Other ways of accessing functions
+# Test that non-ambiguous cases work
+io = IOBuffer()
+precompile(ambig, (Int, Int))
+cfunction(ambig, Int, (Int, Int))
+code_lowered(ambig, (Int, Int))
+code_typed(ambig, (Int, Int))
+code_llvm(io, ambig, (Int, Int))
+code_native(io, ambig, (Int, Int))
+
+# Test that ambiguous fail appropriately
+@test_throws AmbiguousError precompile(ambig, (UInt8, Int))
+# @test_throws AmbiguousError cfunction(ambig, Int, (UInt8, Int))
```

Comment:
This is the only weakness I know of: if you call `cfunction` twice on an ambiguous call, you get a segfault. Haven't yet had time to track this down.